File: android/companion-core/src/main/kotlin/com/xopmc/galaxybridge/core/ADBBindingTranscript.kt
package com.xopmc.galaxybridge.core

import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.security.Signature

/** Canonical proof binding an ADB serial to a paired Android identity. */
object ADBBindingTranscript {
    private val domain = "GalaxyBridge/ADBBinding/v1".encodeToByteArray()

    fun make(
        hostId: String,
        adbSerial: String,
        nonce: ByteArray,
        androidPublicKey: ByteArray,
    ): ByteArray {
        val output = ByteArrayOutputStream()
        DataOutputStream(output).use { writer ->
            listOf(
                domain,
                hostId.encodeToByteArray(),
                adbSerial.encodeToByteArray(),
                nonce,
                androidPublicKey,
            ).forEach { field ->
                writer.writeInt(field.size)
                writer.write(field)
            }
        }
        return output.toByteArray()
    }

    fun verify(
        signatureDer: ByteArray,
        hostId: String,
        adbSerial: String,
        nonce: ByteArray,
        androidPublicKey: ByteArray,
    ): Boolean = runCatching {
        Signature.getInstance("SHA256withECDSA").run {
            initVerify(P256Keys.publicKey(androidPublicKey))
            update(make(hostId, adbSerial, nonce, androidPublicKey))
            verify(signatureDer)
        }
    }.getOrDefault(false)
}

File: Sources/GalaxyBridgeCore/ADBBindingTranscript.swift
import CryptoKit
import Foundation

/// Canonical signed proof that an ADB serial and an already paired companion
/// identity refer to the same physical Android device.
public enum ADBBindingTranscript {
    private static let domain = Data("GalaxyBridge/ADBBinding/v1".utf8)

    public static func make(
        hostID: String,
        adbSerial: String,
        nonce: Data,
        androidPublicKey: Data
    ) -> Data {
        encode([
            domain,
            Data(hostID.utf8),
            Data(adbSerial.utf8),
            nonce,
            androidPublicKey,
        ])
    }

    public static func verify(
        signatureDER: Data,
        hostID: String,
        adbSerial: String,
        nonce: Data,
        androidPublicKey: Data
    ) throws -> Bool {
        let publicKey = try P256.Signing.PublicKey(x963Representation: androidPublicKey)
        let signature = try P256.Signing.ECDSASignature(derRepresentation: signatureDER)
        return publicKey.isValidSignature(
            signature,
            for: make(
                hostID: hostID,
                adbSerial: adbSerial,
                nonce: nonce,
                androidPublicKey: androidPublicKey
            )
        )
    }

    private static func encode(_ fields: [Data]) -> Data {
        var result = Data()
        for field in fields {
            precondition(field.count <= Int(UInt32.max))
            var length = UInt32(field.count).bigEndian
            withUnsafeBytes(of: &length) { result.append(contentsOf: $0) }
            result.append(field)
        }
        return result
    }
}
